elki-core-math/src/main/java/elki/math/statistics/dependence/mcde/MWPTest.java

```diff
@@ -139,53 +142,81 @@ private MWPTest() {
    * @return p-value from two sided Mann-Whitney-U test
    */
-  public double statisticalTest(int start, int end, boolean[] slice, MWPRanking corrected_ranks) {
+  public double statisticalTest(int start, int width, boolean[] slice, MWPRanking corrected_ranks) {
+    final int safeStart = getSafeCut(start, corrected_ranks);
+    final int len = corrected_ranks.len;
+    final int sliceEndSearchStart = safeStart + width > len -1 ? len -1 : safeStart + width;
+    final int safeEnd = getSafeCut(sliceEndSearchStart, corrected_ranks);
+
     double R = 0.0;
     int n1 = 0;
-    for(int j = start; j < end; j++) {
+    for(int j = safeStart; j < safeEnd; j++) {
       if(slice[corrected_ranks.index[j]]) {
         R += corrected_ranks.adjusted[j];
         n1++;
       }
     }
 
-    // Cancel the offset in case the marginal restriction does not start from 0
-    // see "acc-(cutStart*count)" in reference implementation of MWP
-    R -= start * n1;
+    // This is to cancel the offset in case the marginal restriction does not
+    // start from 0 see "acc-(cutStart*count)" is reference implementation of
+    // MWP
+    R -= safeStart * n1;
 
-    final int cutLength = end - start;
+    final int cutLength = safeEnd - safeStart;
     if(n1 == 0 || n1 == cutLength) {
       return 1;
     }
 
     final double U = R - 0.5 * n1 * (n1 - 1);
     final int n2 = cutLength - n1;
-    final long b_end = corrected_ranks.correction[end - 1];
-    final long b_start = start == 0 ? 0 : corrected_ranks.correction[start - 1];
+    final long b_end = corrected_ranks.correction[safeEnd - 1];
+    final long b_start = safeStart == 0 ? 0 : corrected_ranks.correction[safeStart - 1];
     final double correction = (double) (b_end - b_start) / (cutLength * (cutLength - 1));
     final double std = FastMath.sqrt((((double) (n1 * n2)) / 12) * (cutLength + 1 - correction));
-    final double Z = std > 0 ? Math.abs((U - (0.5 * n1 * n2)) / std) : 0.;
+    if(std == 0) { return 0; }
+    final double Z = Math.abs((U - (0.5 * n1 * n2)) / std);
     // Note that this is equivalent to do 1-2*(1-cdf(Z,0,1));
     return NormalDistribution.erf(Z * MathUtil.SQRTHALF);
     // erf(Z / Math.sqrt(2)) is the cdf of the half-normal distribution
   }
 
   /**
-   * Parameterizer, returning the static instance.
-   * 
-   * @author Erich Schubert
+   * Adjusts idx so that it lies not in the middle of a tie by searching the ranks index
+   *
+   * @param idx value to adjust for tie. In MCDE MWP the start and end of a slice
+   * @param ranks the corrected ranks index
+   * @return adjusted idx
    */
-  public static final class Par implements Parameterizer {
-    @Override
-    public MWPTest make() {
-      return STATIC;
+
+  protected int getSafeCut(int idx, MWPRanking ranks){
+    double[] ref = ranks.adjusted;
+
+    int i = 0;
+    while(true){
+      // principle of asking for forgiveness instead of permission, exception gets thrown very rarely
```

Comment:
I don't like this with respect to readability; and AFAIK setting up an exception handler is not exactly for free (it involves managing an exception table). Also the cost of an exception may be much higher than you expect, as it usually involves a full stack trace (there are some ugly hacks with static exceptions, as seen in ELKIs `ParseUtil`, that may be helpful for common exceptions)
See, e.g., https://shipilev.net/blog/2014/exceptional-performance/

Reply:
Done